File: services/alert-service/store.go
package main

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Alert struct {
	ID              string
	OrganizationID  string
	Severity        string
	Status          string
	Title           string
	Description     string
	MachineID       string
	DeviceID        string
	SensorID        string
	FactoryID       string
	TriggeredAt     time.Time
	EscalationLevel int
}

type alertStore struct {
	pool *pgxpool.Pool
}

func newAlertStore(pool *pgxpool.Pool) *alertStore {
	return &alertStore{pool: pool}
}

type createResult string

const (
	resultCreated            createResult = "created"
	resultSuppressedOpen     createResult = "suppressed_open"
	resultSuppressedCooldown createResult = "suppressed_cooldown"
)

// createIfDue implements alert deduplication and cooldown: it refuses to
// create a new alert if one for the same (rule, dedupe scope) is still
// open/unresolved, or if one resolved too recently (within cooldownSeconds)
// — the latter is what stops a flapping condition from re-paging someone
// every time it crosses the threshold.
func (s *alertStore) createIfDue(ctx context.Context, ruleID string, cooldownSeconds int, a Alert, dedupeKey string) (createResult, Alert, error) {
	var lastStatus string
	var lastTriggeredAt time.Time
	err := s.pool.QueryRow(ctx,
		`SELECT status, triggered_at FROM alerts WHERE alert_rule_id = $1 AND dedupe_key = $2 ORDER BY triggered_at DESC LIMIT 1`,
		ruleID, dedupeKey,
	).Scan(&lastStatus, &lastTriggeredAt)

	if err != nil && !errors.Is(err, pgx.ErrNoRows) {
		return "", Alert{}, fmt.Errorf("check existing alert: %w", err)
	}
	if err == nil {
		if lastStatus != "RESOLVED" {
			return resultSuppressedOpen, Alert{}, nil
		}
		if time.Since(lastTriggeredAt) < time.Duration(cooldownSeconds)*time.Second {
			return resultSuppressedCooldown, Alert{}, nil
		}
	}

	var id string
	var triggeredAt time.Time
	err = s.pool.QueryRow(ctx,
		`INSERT INTO alerts (organization_id, alert_rule_id, factory_id, machine_id, device_id, sensor_id, severity, status, title, description, dedupe_key)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, 'OPEN', $8, $9, $10)
		 ON CONFLICT (alert_rule_id, dedupe_key) WHERE status = 'OPEN' DO NOTHING
		 RETURNING id, triggered_at`,
		a.OrganizationID, ruleID, nullIfEmpty(a.FactoryID), nullIfEmpty(a.MachineID), nullIfEmpty(a.DeviceID), nullIfEmpty(a.SensorID),
		a.Severity, a.Title, a.Description, dedupeKey,
	).Scan(&id, &triggeredAt)

	if errors.Is(err, pgx.ErrNoRows) {
		// Lost a race with a concurrent insert for the same open alert.
		return resultSuppressedOpen, Alert{}, nil
	}
	if err != nil {
		return "", Alert{}, fmt.Errorf("insert alert: %w", err)
	}

	a.ID = id
	a.Status = "OPEN"
	a.TriggeredAt = triggeredAt
	return resultCreated, a, nil
}

// dueForEscalation returns OPEN alerts that have gone unacknowledged past
// afterSeconds since their last escalation (or since triggering, if never
// escalated), for the periodic escalation sweep.
func (s *alertStore) dueForEscalation(ctx context.Context, afterSeconds int) ([]Alert, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT id, organization_id, severity, title, description,
		       COALESCE(machine_id::text, ''), COALESCE(device_id::text, ''),
		       triggered_at, escalation_level
		FROM alerts
		WHERE status = 'OPEN'
		  AND escalation_level < 2
		  AND COALESCE(last_escalated_at, triggered_at) < now() - make_interval(secs => $1)
	`, afterSeconds)
	if err != nil {
		return nil, fmt.Errorf("query alerts due for escalation: %w", err)
	}
	defer rows.Close()

	var alerts []Alert
	for rows.Next() {
		var a Alert
		if err := rows.Scan(&a.ID, &a.OrganizationID, &a.Severity, &a.Title, &a.Description,
			&a.MachineID, &a.DeviceID, &a.TriggeredAt, &a.EscalationLevel); err != nil {
			return nil, fmt.Errorf("scan escalation candidate: %w", err)
		}
		alerts = append(alerts, a)
	}
	return alerts, rows.Err()
}

var severityLadder = []string{"WARNING", "HIGH", "CRITICAL"}

func nextSeverity(current string) string {
	for i, s := range severityLadder {
		if s == current && i+1 < len(severityLadder) {
			return severityLadder[i+1]
		}
	}
	return current
}

// escalate bumps an alert's severity and escalation_level, returning the
// new severity for the re-notification.
func (s *alertStore) escalate(ctx context.Context, alertID, newSeverity string) error {
	_, err := s.pool.Exec(ctx,
		`UPDATE alerts SET severity = $1, escalation_level = escalation_level + 1, last_escalated_at = now(), updated_at = now() WHERE id = $2`,
		newSeverity, alertID,
	)
	if err != nil {
		return fmt.Errorf("escalate alert %s: %w", alertID, err)
	}
	return nil
}

func nullIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

File: services/alert-service/store.py
"""Alert persistence, deduplication, cooldown, and escalation, the Python
port of store.go."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from psycopg_pool import ConnectionPool

CREATED = "created"
SUPPRESSED_OPEN = "suppressed_open"
SUPPRESSED_COOLDOWN = "suppressed_cooldown"

_SEVERITY_LADDER = ["WARNING", "HIGH", "CRITICAL"]


@dataclass
class Alert:
    organization_id: str
    severity: str
    factory_id: str = ""
    machine_id: str = ""
    device_id: str = ""
    sensor_id: str = ""
    title: str = ""
    description: str = ""
    id: str = ""
    status: str = ""
    triggered_at: datetime | None = None
    escalation_level: int = 0


class AlertStore:
    def __init__(self, pool: ConnectionPool):
        self._pool = pool

    def create_if_due(self, rule_id: str, cooldown_seconds: int, alert: Alert, dedupe_key: str) -> tuple[str, Alert | None]:
        """Implements alert deduplication and cooldown: refuses to create a
        new alert if one for the same (rule, dedupe scope) is still
        open/unresolved, or if one resolved too recently (within
        cooldown_seconds) — the latter is what stops a flapping condition
        from re-paging someone every time it crosses the threshold."""
        with self._pool.connection() as conn:
            row = conn.execute(
                "SELECT status, triggered_at FROM alerts WHERE alert_rule_id = %s AND dedupe_key = %s ORDER BY triggered_at DESC LIMIT 1",
                (rule_id, dedupe_key),
            ).fetchone()

            if row is not None:
                last_status, last_triggered_at = row
                if last_status != "RESOLVED":
                    return SUPPRESSED_OPEN, None
                if datetime.now(timezone.utc) - last_triggered_at < timedelta(seconds=cooldown_seconds):
                    return SUPPRESSED_COOLDOWN, None

            inserted = conn.execute(
                """
                INSERT INTO alerts (organization_id, alert_rule_id, factory_id, machine_id, device_id, sensor_id, severity, status, title, description, dedupe_key)
                VALUES (%s, %s, %s, %s, %s, %s, %s, 'OPEN', %s, %s, %s)
                ON CONFLICT (alert_rule_id, dedupe_key) WHERE status = 'OPEN' DO NOTHING
                RETURNING id, triggered_at
                """,
                (
                    alert.organization_id, rule_id, _null_if_empty(alert.factory_id), _null_if_empty(alert.machine_id),
                    _null_if_empty(alert.device_id), _null_if_empty(alert.sensor_id),
                    alert.severity, alert.title, alert.description, dedupe_key,
                ),
            ).fetchone()

        if inserted is None:
            # Lost a race with a concurrent insert for the same open alert.
            return SUPPRESSED_OPEN, None

        # id comes back as a uuid.UUID (not str) since the RETURNING clause
        # doesn't cast it -- same psycopg3 gotcha as elsewhere in this port.
        alert.id, alert.triggered_at = str(inserted[0]), inserted[1]
        alert.status = "OPEN"
        return CREATED, alert

    def due_for_escalation(self, after_seconds: int) -> list[Alert]:
        """Returns OPEN alerts that have gone unacknowledged past
        after_seconds since their last escalation (or since triggering, if
        never escalated), for the periodic escalation sweep."""
        with self._pool.connection() as conn:
            rows = conn.execute(
                """
                SELECT id, organization_id, severity, title, description,
                       COALESCE(machine_id::text, ''), COALESCE(device_id::text, ''),
                       triggered_at, escalation_level
                FROM alerts
                WHERE status = 'OPEN'
                  AND escalation_level < 2
                  AND COALESCE(last_escalated_at, triggered_at) < now() - make_interval(secs => %s)
                """,
                (after_seconds,),
            ).fetchall()

        return [
            Alert(
                id=str(r[0]), organization_id=str(r[1]), severity=r[2], title=r[3], description=r[4],
                machine_id=r[5], device_id=r[6], triggered_at=r[7], escalation_level=r[8],
            )
            for r in rows
        ]

    def escalate(self, alert_id: str, new_severity: str) -> None:
        """Bumps an alert's severity and escalation_level, for the caller
        to use in its re-notification."""
        with self._pool.connection() as conn:
            conn.execute(
                "UPDATE alerts SET severity = %s, escalation_level = escalation_level + 1, last_escalated_at = now(), updated_at = now() WHERE id = %s",
                (new_severity, alert_id),
            )


def next_severity(current: str) -> str:
    for i, s in enumerate(_SEVERITY_LADDER):
        if s == current and i + 1 < len(_SEVERITY_LADDER):
            return _SEVERITY_LADDER[i + 1]
    return current


def _null_if_empty(s: str) -> str | None:
    return s if s else None
